Indentation rework

Triton171 opened the pull request:
I reworked the indentation system to remove some of it's limitations and make it easier to extend. This should close #114, #1523 and #1440.

Things I want to implement in followup PRs:
- Specify nodes to be extended to blank lines (and a list of "return" nodes, that interrupt this behavior, as it's done in nvim-treesitter). This is required to correctly indent languages such as python where treesitter nodes may be too short
- Specify nodes with a custom prefix (e.g. to automatically make a new line a comment)
- Specify nodes inside which no treesitter indentation is used (e.g. for multiline string literals)
- Specify nodes which should be aligned with each other
- Add indentation tests for other languages

sudormrfbin commented:
I would recommend splitting all the TODOs to separate small PRs since it will be easier to review and keep track of.

Triton171 commented:
Thanks @sudormrfbin, splitting it up is a good idea. 
This PR now just implements the core rework and migration of all existing indent queries. I've added a few rules to the Rust, C and C++ queries that show what the new features can be used for. Additionally, I've added one missing rule to the Go indent query to fix #1523.

pickfire commented on helix-core/src/syntax.rs:
```diff
@@ -106,15 +106,88 @@ pub struct IndentationConfiguration {
+    // A node given by a list of characteristics which must all be fulfilled in order to match
+    ComplexNode {
+        kind: Option<String>,
+        kind_not_in: Option<Vec<String>>,
+        parent_kind_in: Option<Vec<String>>,
+        field_name_in: Option<Vec<String>>,
```
Why not just `Vec<String>`?

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,291 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
-/// This is to identify the column where this node (e.g., an HTML closing tag) ends.
-fn get_highest_syntax_node_at_bytepos(syntax: &Syntax, pos: usize) -> Option<Node> {
-    let tree = syntax.tree();
+struct AddedIndent {
```
Can you please add some comments on these structs? I think at least the core stuff should be documented.

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,291 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
-        Some(node) => node,
-        None => return None,
-    };
+struct IndentResult {
+    indent: i32,
+}
```
Can you please add a comment here saying that indentation is in terms of levels and not spaces? Also, I wonder how well this will work for python indention for this:
```python
def hello(a,
          b,
          c):
```

given that the indentation depends on the previous line, if there are no items there.

Triton171 replied:
Regarding python, I have this on my radar, I'm planning to add something like an `align` specification. I still have to think about how to make it as intuitive and general as possible. It's kinda difficult to define what to make the "anchor" for the alignment. For simple cases (like aligning the parameters of a function) it should not be to complicated though, since treesitter does all the heavy lifting.

Triton171 commented:
I added some more detailed documentation which should hopefully explain the general idea. @archseer let me know if anything is unclear.

pickfire commented on helix-core/src/syntax.rs:
```diff
@@ -106,15 +106,92 @@ pub struct IndentationConfiguration {
+    ComplexNode {
+        kind: Option<String>,
+        #[serde(default)]
+        #[serde(skip_serializing_if = "Vec::is_empty")]
```
Why do we need to skip serializing if empty?

Triton171 replied:
I'm actually not sure why the `Configuration` struct needs to be serializable (Helix also compiles without that implementation), so I didn't know how the `Serialize` impl should behave. Let me know what behavior we want here (or if we even need the impl at all).

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+        if added.indent && !added.outdent {
+            self.indent += 1;
+        } else if added.outdent && !added.indent {
+            self.indent = self.indent.saturating_sub(1);
```
Is this possible to underflow?

Triton171 replied:
Nope, just a remnant of when indent was an unsigned integer...

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    fn as_string(&self, indent_style: &IndentStyle) -> String {
+        indent_style
+            .as_str()
+            .repeat(std::cmp::max(self.indent, 0) as usize)
```
```suggestion
            .repeat(0.max(self.indent as usize))
```

Triton171 replied:
Definitely looks nicer. I think we still need to take the maximum before casting, otherwise a negative value for `self.indent` (which could happen depending on the indent query) will lead to very weird results

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+                || (new_line && node.start_byte() >= byte_pos && prev.start_byte() < byte_pos);
+            first_in_line.push(Some(first));
+        } else {
+            // Nodes that have no previous siblings are first in their line iff their parent is
```
iff?

Triton171 replied:
It's an abbreviation for "if and only if". I'm actually not sure how common it is outside of mathematics so I removed it.

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
     }
 
-    Some(node)
+    let mut result = Vec::with_capacity(first_in_line.len());
+    let mut parent_is_first = true; // The root node is by definition the first node in its line
+    for first in first_in_line.into_iter().rev() {
+        if let Some(first) = first {
+            result.push(first);
+            parent_is_first = first;
+        } else {
+            result.push(parent_is_first);
+        }
+    }
+    result
```
Why not just do collect rather than manually pushing it into vec?

Triton171 replied:
How would we construct the iterator to collect from? Maybe there's some iterator function I'm missing but because we need the external state `parent_is_first`, this doesn't seem like it's nicely doable with iterators.

pickfire replied:
```suggestion
    // The root node is by definition the first node in its line
    first_in_line
        .into_iter()
        .rev()
        .scan(true, |parent_is_first, first| {
            Some(if let Some(first) = first {
                *parent_is_first = first;
                first
            } else {
                *parent_is_first
            })
        })
        .collect()
```

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+            return true;
+        }
+    }
+    for unnamed_node in scope.iter().take_while(|n| n.kind().is_none()) {
+        if matches(unnamed_node, node, cursor) {
+            return true;
+        }
+    }
+    false
```
This can be simplified with `any`

Triton171 replied:
Yeah, `any` is nice. I updated the part with the named nodes above as well to use it.

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    let byte_pos = text.char_to_byte(pos);
+    let mut node = match get_lowest_node(syntax.tree().root_node(), query, byte_pos) {
+        Some(n) => n,
+        None => {
+            return None;
         }
-        if query.indent.contains(node_kind) {
-            increment_from_line += 1;
+    };
```
This can just use `?`

Triton171 replied:
Nice, I didn't know the `?` operator could be used with `Option` too.

pickfire commented:
Also, please fix the lints.

pickfire commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +188,328 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+                        }
+                        break;
+                    }
+                    debug_assert!(cursor.goto_next_sibling());
```
Won't this only be in debug and result in infinite loop here for release build?

pickfire commented:
My review ends here. I think it's better for @archseer to check since I don't haven't done indent much.

archseer commented on runtime/queries/rust/indents.toml:
```diff
@@ -1,4 +1,9 @@
-indent = [
+[indent]
+all = [
+ { parent_kind_in = ["assignment_expression", "compound_assignment_expr"], field_name_in = ["right"]},
+ { parent_kind_in = ["let_declaration"], field_name_in = ["value"] }
+]
+tail = [
```
I've read through both the examples the docs and the original issue, but I don't understand the need for `all` vs `tail`.

The point is to make:

```rust
let (
    a_long_variable_name_in_this_tuple,
    b_long_variable_name_in_this_tuple,
    c_long_variable_name_in_this_tuple,
    d_long_variable_name_in_this_tuple,
    e_long_variable_name_in_this_tuple,
): (usize, usize, usize, usize, usize) = //<- cursor here
```

`<Enter>` produce indentation level == 1 here?

Triton171 replied:
Yes, basically. `indent.tail` produces the behavior of the old `indent` option (indenting everything except for the first line of a node). `indent.all` indents the whole node including the first line, similar to how the old `outdent` behaves. Here we need it because the whole RHS of a let declaration should be indented, including the first line. Additionally, `indent` and `outdent` are symmetric in the new version, so we don't need any special cases for them and for everything else that the indent code might handle in the future (comments, alignment, etc.).

Your example doesn't work because of a limitation of tree-sitter: In that situation there isn't a value for the `let_declaration` yet, so we don't know that we should indent it. The best solution here is probably an optional auto-indent feature that updates the indent as soon as tree-sitter parses the code correctly.
Pressing `<Enter>` in this situation does produce the correct indent (1):
```rust
let (
    a_long_variable_name_in_this_tuple,
    b_long_variable_name_in_this_tuple,
    c_long_variable_name_in_this_tuple,
    d_long_variable_name_in_this_tuple,
    e_long_variable_name_in_this_tuple,
): (usize, usize, usize, usize, usize) = |(0,0,0,0,0);
// Cursor at | in the line above
```

Triton171 commented:
I just had a closer look at the Treesitter query documentation and code and realized I misunderstood the API. Setting the byte range for a query doesn't just find all matches inside that range, it finds all matches that _intersect_ the range. This means that we can query a very small range (usually a couple of bytes should be enough) and still get all the relevant matches for indentation.

Unless there's some disadvantage that I'm not seeing, I'll change this PR to use treesitter queries instead of the custom TOML format, as was suggested in #114. I still think we shouldn't just try to copy the indent queries from nvim-treesitter because they don't support stuff like custom prefixes (e.g. for comments or markdown lists). Maybe I'm wrong though, so suggestions/discussions are welcome.

Triton171 commented:
Yes it will. I also tested it with the following code:
```rust
fn test() {
    do_smth();
    fn nested() {
        let a = 0; // <- Run the query on this line
    }
    do_more();
}
```

Running the query `(_) @capture` on (a subset of) the fourth line yields the following captured node kinds:
- source_file
- function_item
- block
- function_item
- block
- let_declaration

Basically, it will return match that intersects the given byte range. This includes nodes that contain the byte range completely. [This](https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/query.c#L3315) is the relevant part of the tree-sitter code.

archseer commented:
Hmm okay, if you can get it working with tree-sitter queries then that would be a big improvement 👍🏻

archseer commented on runtime/queries/rust/indents.scm:
```diff
@@ -0,0 +1,38 @@
+
+(assignment_expression
+  right: (_) @indent
+  (#set! "scope" "all"))
```
This looks quite nice, a reasonable default but then being able to override the behavior with a #set! 👍

Triton171 commented:
I finished porting the code to use tree-sitter queries. I tested that all new queries to make sure they load correctly (some had invalid/outdated nodes which I removed). There's now also a section for indent queries in the book which is probably the best place to start to understand the new system.

Triton171 commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+                            Some("all") => IndentScope::All,
+                            Some("tail") => IndentScope::Tail,
+                            Some(s) => {
+                                panic!("Invalid indent query: Unknown value for \"scope\" property (\"{}\")", s);
```
Is this the correct way to handle an invalid query? I feel like at least in debug mode it should not fail silently, because otherwise debugging indent queries might become a bit frustrating. Maybe panic in debug mode and ignore in release?

archseer replied:
I agree, we could make a macro that would panic in debug, log in release. 

But it's fine to do that in a follow-up since we don't do this for other query types right now either.

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+            for property in query.property_settings(m.pattern_index) {
+                match property.key.as_ref() {
+                    "scope" => {
+                        indent_capture.scope = match property.value.as_ref().map(|s| s.as_ref()) {
```
`&property.value.as_deref()` maybe?

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+                    }
+                }
+            }
+            match indent_captures.entry(capture.node.id()) {
+                Entry::Occupied(mut e) => {
+                    e.get_mut().push(indent_capture);
+                }
+                Entry::Vacant(e) => {
+                    e.insert(vec![indent_capture]);
+                }
```
`indent_captures.entry(capture.node.id()).or_default().push(indent_capture)`

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
         if let Some(parent) = node.parent() {
+            let mut node_line = node.start_position().row;
+            let mut parent_line = parent.start_position().row;
+            if node.start_position().row == line && new_line {
```
```suggestion
            if node_line.row == line && new_line {
```

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
-            node = parent
+/// Computes for node and all ancestors whether they are the first node on their line.
+/// The first entry in the return value represents the root node, the last one the node itself
+fn get_first_in_line(mut node: Node, byte_pos: usize, new_line: bool) -> Vec<bool> {
```
Isn't it easier to get the first non-whitespace char position in the line (we have a function for that already), then use that with `descendant_for_byte_range`?

Triton171 replied:
Yes it would be easier but probably not correct for all grammars. For example we might at some point have a grammar for literate haskell. Because code is marked with a `>` at the beginning of the line, just taking the first non-whitespace char would always (incorrectly) produce the first char.

archseer replied:
But wouldn't `get_first_in_line` also match this first character?

Triton171 replied:
It shouldn't because a reasonable grammar for literate haskell would probably not have `>` as a node in that context and instead ignore it. This enables us to ignore more text than just whitespace if the language requires it.  It might also work better for language injection which we probably want to implement at some point for indent queries.

If you think that this is too specific an issue I can also replace this with simpler approach of just using whitespace.

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    fn new() -> Self {
+        IndentResult { indent: 0 }
+    }
+    /// Add the given [AddedIndent] to the [IndentResult].
+    /// The [AddedIndent] should be the combination of all the added indents for one line.
+    fn add(&mut self, added: &AddedIndent) {
+        if added.indent && !added.outdent {
+            self.indent += 1;
+        } else if added.outdent && !added.indent {
+            self.indent -= 1;
+        }
+    }
+    fn as_string(&self, indent_style: &IndentStyle) -> String {
+        indent_style.as_str().repeat(0.max(self.indent) as usize)
+    }
```
This isn't safe and will panic if the indent ends up being below zero. I'd rather maintain an indent and outdent usize count, then get the difference between the two. (We can return 0 if outdent > indent and log an error)

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+/// The indent that is added for a single line.
+/// This is different from the total indent ([IndentResult]) because multiple indents/outdents on the same line don't stack.
+/// It should be constructed by successively adding the relevant indent captures.
+struct AddedIndent {
+    indent: bool,
+    outdent: bool,
+}
+impl AddedIndent {
+    /// Returns the [AddedIndent] for a line without any indent definitions
+    fn new() -> Self {
+        AddedIndent {
+            indent: false,
+            outdent: false,
+        }
+    }
+    /// Adds an indent capture to this indent.
+    fn add_capture(&mut self, capture: &IndentCaptureType) {
+        match capture {
+            IndentCaptureType::Indent => {
+                self.indent = true;
+            }
+            IndentCaptureType::Outdent => {
+                self.outdent = true;
+            }
+        }
+    }
+}
```
With the proposal above you could then drop AddedIndent altogether and use a single `Indent` type

`Indent { indent: 1, outdent: 0 }` etc

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+        }
+    }
+    /// Adds an indent capture to this indent.
+    fn add_capture(&mut self, capture: &IndentCaptureType) {
```
No reason to pass by ref, just derive `Copy` on simple enums and structs.

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+}
+impl IndentCaptureType {
+    fn default_scope(&self) -> IndentScope {
+        match *self {
```
And with Copy derived then this no longer needs to `*` I think.

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    let mut indent_for_line = AddedIndent::new();
+    let mut indent_for_line_below = AddedIndent::new();
+    loop {
+        let is_first = *first_in_line.last().unwrap();
```
This unwrap doesn't seem to be safe, I'd `?` to return `None`?

Triton171 replied:
It should be safe: `first_in_line` contains an entry for each ancestor of `node` (and for the `node` itself). This is exactly what we traverse over in that loop, removing one element from `first_in_line` after each iteration.

archseer replied:
Can you add a comment for that?

archseer commented on helix-core/src/indent.rs:
```diff
@@ -346,6 +578,32 @@ mod test {
 
         let does_indentation_work = 1;
 
+        let mut really_long_variable_name_using_up_the_line =
```
If calculation is good enough to match `cargo fmt` yet, we can `include_str!()` `.rs` files in this crate or `helix-term/src/commands.rs` and calculate indentation for the entire file. I was originally doing that but the formatter wasn't able to match `cargo fmt` in all cases.

Triton171 replied:
I've thought about this as well, the current solution with an inline string isn't ideal. Could we maybe have a `test-indent` folder somewhere for these test files? That way we could also test languages other than rust (some future features like `align` don't really apply to rust). For rust, we could still just add a symlink.

archseer replied:
Agreed!

Triton171 commented:
I've improved the rust query a bit and implemented your suggestions. I also tried running the indentation test on `helix-term/src/commands.rs`. It now matches the indentation for the first ~3500 lines, then it fails because the indentation system currently can't align stuff (which is required when the parameter list of a closure spans multiple lines). I'll implement that in a follow-up PR, once we can handle it, `commands.rs` should be a good test for rust indentation.

archseer commented:
Looks good, I'm taking a deeper look at indent.rs and I'm test driving this branch for a few days.

archseer commented:
Regarding cargo fmt: it was probably for code inside a macro? It seems to avoid any formatting inside them

archseer commented on helix-core/src/indent.rs:
```diff
@@ -328,97 +622,28 @@ mod test {
+        use crate::syntax::{Configuration, IndentationConfiguration, Loader};
         use once_cell::sync::OnceCell;
+        use std::path::PathBuf;
+        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        path.push("test-indent");
+        path.push(file_name);
```
Let's put the fixtures under `helix-core/test/data/indent` or `helix-core/test/fixtures/indent`. The test cases themselves could also be under `test/`

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    /// Add some other [IndentResult] to this.
+    /// The added indent should be the total added indent from one line
+    fn add_line(&mut self, added: &IndentResult) {
+        if added.indent > 0 && added.outdent == 0 {
+            self.indent += 1;
+        } else if added.outdent > 0 && added.indent == 0 {
+            self.outdent += 1;
+        }
```
This seems odd to me, can't `indent` and `outdent` both be `> 0`?

archseer replied:
You're also not using the number from `added` and always use `1`.

Triton171 replied:
This is because of how the indentation that's added from a single line works: We can always have at most one indent or at most one outdent (because the indentation never increases by multiple levels on a single line). Here, `added.indent` and `added.outdent` will always be 1 or 0 because `added` is constructed using the function `add_capture`. I just check for `>0` instead for `==1` to prevent weird errors in case this invariant no longer holds in the future.

If there are both `indent` and `outdent` nodes on the same line, we ignore all of them, that's why there is no third case for when `added.indent > 0 && added.outdent > 0`. This is required to fix #1523.

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+/// This is usually constructed in one of 2 ways:
+/// - Successively add indent captures to get the (added) indent from a single line
+/// - Successively add the indent results for each line
+struct IndentResult {
```
Rather than `IndentResult` how about calling this `Indentation`?

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    outdent: usize,
+}
+impl IndentResult {
+    fn new() -> Self {
+        IndentResult {
+            indent: 0,
+            outdent: 0,
+        }
+    }
```
Could derive `Default` then use `Self::default()`

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    new_line_break: Option<(usize, usize)>,
+) -> HashMap<usize, Vec<IndentCapture>> {
+    // TODO Maybe reuse this cursor?
+    let mut cursor = QueryCursor::new();
```
It's a bit of a mess but you can pull a cursor from the pool here: https://github.com/helix-editor/helix/blob/738ea7acd9a01a259f66f61044a9c895f034e384/helix-core/src/syntax.rs#L607-L609

archseer commented on helix-core/src/indent.rs:
```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+        }) {
+            continue;
+        }
+        for capture in m.captures {
```
Is there a reasonable guess for capacity on `indent_captures` we could use? Either `::with_capacity()` or by calling `reserve()` inside this loop

Triton171 replied:
It corresponds to the number of times this node is captured with an IndentCapture (currently `@indent` and `@outdent`. This should be 1 in most cases (because we don't create the `Vec` if it's 0), I added a `with_capacity` to make that a bit more explicit.

the-mikedavis commented on runtime/queries/javascript/indents.scm:
```diff
@@ -0,0 +1,22 @@
+] @indent
+
+[
+ "}"
+ "]"
+ ")"
```
```suggestion
  "}"
  "]"
  ")"
```

Sorry this is a bit of a nit. I see the single-space indent in some other indent files as well:

- dart
- go
- llvm
- nix
- scala
- zig

At some point I'd really like to write a formatter for tree-sitter queries to make this all moot 😄

Triton171 replied:
A formatter would be pretty nice. I must've missed this while converting all the indent queries, thanks for pointing out :)

the-mikedavis reviewed (commented):
this is very exciting! 😄

archseer commented on helix-core/src/syntax.rs:
```diff
@@ -85,17 +85,17 @@ pub struct LanguageConfiguration {
     //
     #[serde(skip)]
-    pub(crate) highlight_config: OnceCell<Option<Arc<HighlightConfiguration>>>,
+    pub highlight_config: OnceCell<Option<Arc<HighlightConfiguration>>>,
```
Were these made `pub` because of usage in tests? Why not use the associated methods `highlight_config()` and `indent_query()`?

archseer replied:
Ah, it's because `LanguageConfiguration` is constructed directly in the test. Let's instead use the loader with a custom `languages.toml`? Can reuse this one: https://github.com/helix-editor/helix/blob/ce8311727f32975dc1c1030c676ffc141f63b851/.github/workflows/languages.toml from #1659

archseer commented:
Tested over the week and it works great! Once the last set of comments is resolved this is ready for merge 👍🏻

Triton171 commented:
Nice to hear that! I probably won't have time to work on it in the next week or so but I definitely want to finish it (and work on same of the follow-up features) when I can.

max-privatevoid commented:
I wrote a basic [test suite](https://github.com/max-privatevoid/helix-tests) (currently specific to Nix) to test this PR. The following test cases currently fail:
```diff
TESTING: let-in-attr
*** work/let-in-attr.nix	2022-03-05 21:29:07.460718670 +0100
--- output/let-in-attr.nix	2022-03-05 21:28:24.453336376 +0100
***************
*** 1,7 ****
  let
    a = 3;
  in {
    b = [
!       /*DONE*/
    ];
  }
--- 1,7 ----
  let
    a = 3;
  in {
    b = [
!     /*DONE*/
    ];
  }
TESTING: let
*** work/let.nix	2022-03-05 21:29:08.416727170 +0100
--- output/let.nix	2022-03-05 19:31:23.883676561 +0100
***************
*** 1,2 ****
  let
! /*DONE*/
--- 1,2 ----
  let
!   /*DONE*/
```
The `/*DONE*/` marker represents where the cursor would end up at when adding a new line from the line above.


---

In `let-in-attr.nix`, Helix seems to pick which indentation level to use in relation to the layout of the entire file rather than the context immediately around the current line. Sometimes, "incorrect" indentations may be preferred for stylistic reasons, such as:
```nix
let
  value = 1;
in {
  inherit value;
}
```
vs the variant Helix would want:
```nix
let
  value = 1;
in {
    inherit value;
  }
```
Nix formatters such as [alejandra](https://github.com/kamadorueda/alejandra) would transform the latter into the former.

This also happens with other languages, such as JSON. Here, the indentation before "a" was incorrectly removed. Helix places the cursor and closing square bracket on the correct level, while carrying over the incorrect indentation would look more consistent.
```json
{
"a": [
    | <- cursor ends up here
  ]
}
```
```json
{
"a": [
  | <- cursor should end up here
]
}
```
---

The case of `let.nix` seems to be a matter of dealing with indents while syntax errors are present due to incomplete statements. A `let` with a missing `in` results in no indentation on the next line. If the `in` is added here and a new line is then created after `let`, it gets indented by one level.

As for why a different behavior would be desirable: Writing `let`, then some declarations, then `in` is a more natural flow than writing `let` and `in`, then going back up to write the declarations inbetween the two keywords.

Triton171 commented:
Thanks for bringing up these issues, even though some of them will probably be a bit complicated to fix.

The incorrect indentation for `let ... in` expressions should be fixable with the current system. The issue is that both the `let` as well as the `parenthesized` node produce a level of indentation. This could be fixed by replacing the `@indent` for `let` expressions with an `@indent.all` for any node whose parent is `let` and that is not `parenthesized` or `in` (you can use `#not-kind-eq`) for that. I don't know when/if I'll get to improving Nix indentation but I'm happy to help in case you want to open a PR (once this is merged).
Once it works, having Nix tests in `helix-core/tests/indent.rs` would also be great.

The fact that indentation is calculated globally can in my opinion be seen both as an advantage and a disadvantage: It manages to correctly compute indentation even when the current indentation is wrong in some places. It just becomes annoying when this "wrong" indentation is actually intentional. We could have an option where we don't apply indent from a tree-sitter node when the indent doesn't exist in the file. For me this isn't super high-priority though.

For the indentation of incomplete code, i can think of 2 possible solutions:

- Explicitly handle common cases in `indents.scm` by writing query patterns that contain errors (I've seen this approach in nvim-treesitter). This should already be possible after this PR.
- Add an (optional) auto-indent feature. This would run on every keystroke and update the code when the correct indent changes. This would need to be smart in order to prevent "correcting" the indent after the user has explicitly set it. Also, it should be enabled on a per-language basis for languages with well-tested indent queries. This would have 2 advantages over the previous approach but would take more effort initially:
  - It doesn't add complexity to the indent query for each languages, it only requires them to be reasonably accurate.
  - It also works in ambiguous situations. For example in C, the line after `if (condition)` should only be indented, if it isn't the beginning of a `compound_statement` (i.e. if the next character is not `{`).

archseer reviewed (approved):
Looks good to me now, stellar work @Triton171! 🎉
